src/main/java/com/google/devtools/build/lib/runtime/AuthHeadersProvider.java

```diff
@@ -0,0 +1,48 @@
+package com.google.devtools.build.lib.runtime;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Generic interface to provide authentication headers for http/grpc requests
+ * to bazel modules.
+ *
+ * <p>Implementations need to be thread-safe.
+ */
+@ThreadSafe
+public interface AuthHeadersProvider {
+
+  /**
+   * Returns the type of authentication mechanism used i.e. oauth.
+   */
+  String getType();
+
+  /**
+   * Returns request headers necessary for authentication to be added to the http/grpc request.
+   *
+   * <p>For auth tokens with a TTL attached an implementation is expected to transparently
+   * refresh the tokens and to always return headers with sufficient TTL left to complete
+   * the request.
+   */
+  Map<String, List<String>> getRequestHeaders(URI uri) throws IOException;
+
+  /**
+   * Refreshes the authentication credentials i.e. if an access token expired.
+   *
+   * <p>This method should be called by clients if a request fails due to expired authentication
```

Comment:
if the implementation "is expected to transparently refresh tokens", why does a client need to call refresh() explicitly?

Reply:
That's because popular authentication methods typically create temporary tokens with an associated TTL value. This has an inherent race, as some time passes between the client constructing the request and the server receiving and validating it. While this failure condition is rare it's nevertheless important that clients handle it in order to be robust.